File: backend/models/book.model.js
const mongoose = require("mongoose");

const bookSchema = new mongoose.Schema({
  title: { type: String, required: true, text: true },
  author: { type: String, required: true, text: true },
  genre: { type: String, required: true, text: true },
  isbn: { type: String, required: true, unique: true },
  description: { type: String },
  coverImagePath: { type: String, default: "" },
  status: {
    type: String,
    enum: ["available", "borrowed", "reserved"],
    default: "available",
  },
  addedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
  currentBorrower: { type: mongoose.Schema.Types.ObjectId, ref: "User", default: null },
  reservationCount: { type: Number, default: 0 },
  nextAvailableDate: { type: Date, default: null },
  borrowCount: { type: Number, default: 0 },
}, { timestamps: true });


// Enable text search on title, author, and genre
bookSchema.index({ title: "text", author: "text", genre: "text" });

module.exports = mongoose.model("Book", bookSchema);

File: lib/models/book_model.dart
class BookModel {
  final String? id; // made nullable to support adding
  final String title;
  final String? author;
  final String? genre;
  final String? isbn;
  final String? description;
  final bool available;
  final String? status;

  BookModel({
    this.id,
    required this.title,
    this.author,
    this.genre,
    this.isbn,
    this.description,
    this.available = true,
    this.status,
  });

  factory BookModel.fromJson(Map<String, dynamic> json) {
    return BookModel(
      id: json['_id'] ?? json['id'],
      title: json['title'] ?? '',
      author: json['author'],
      genre: json['genre'],
      isbn: json['isbn'],
      description: json['description'],
      available: json['available'] ?? true,
      status: json['status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'title': title,
      'author': author,
      'genre': genre,
      'isbn': isbn,
      'description': description,
      'available': available,
      if (status != null) 'status': status,
    };
  }

  bool get isAvailable => status == 'available';
}
